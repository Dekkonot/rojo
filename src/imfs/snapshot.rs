// This file is non-critical and used for testing, so it's okay if it's unused.
#![allow(unused)]

use std::collections::HashMap;

#[derive(Debug, Clone)]
pub enum ImfsSnapshot {
    File(FileSnapshot),
    Directory(DirectorySnapshot),
}

impl ImfsSnapshot {
    /// Create a new file ImfsSnapshot with the given contents.
    pub fn file(contents: impl Into<Vec<u8>>) -> ImfsSnapshot {
        ImfsSnapshot::File(FileSnapshot {
            contents: contents.into(),
        })
    }

    /// Create a new directory ImfsSnapshot with the given children.
    pub fn dir<S: Into<String>>(children: HashMap<S, ImfsSnapshot>) -> ImfsSnapshot {
        let children = children.into_iter().map(|(k, v)| (k.into(), v)).collect();

        ImfsSnapshot::Directory(DirectorySnapshot { children })
    }

    pub fn empty_dir() -> ImfsSnapshot {
        ImfsSnapshot::Directory(DirectorySnapshot {
            children: Default::default(),
        })
    }
}

#[derive(Debug, Clone)]
pub struct FileSnapshot {
    pub contents: Vec<u8>,
}

#[derive(Debug, Clone)]
pub struct DirectorySnapshot {
    pub children: HashMap<String, ImfsSnapshot>,
}
