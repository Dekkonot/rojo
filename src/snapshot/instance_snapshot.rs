//! Defines the structure of an instance snapshot.

use std::{borrow::Cow, collections::HashMap};

use rbx_dom_weak::{RbxId, RbxTree, RbxValue};
use serde::{Deserialize, Serialize};

use super::InstanceMetadata;

/// A lightweight description of what an instance should look like. Attempts to
/// be somewhat memory efficient by borrowing from its source data, indicated by
/// the lifetime parameter `'source`.
///
// Possible future improvements:
// - Use refcounted/interned strings
// - Replace use of RbxValue with a sum of RbxValue + borrowed value
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InstanceSnapshot<'source> {
    /// A temporary ID applied to the snapshot that's used for Ref properties.
    pub snapshot_id: Option<RbxId>,

    /// Rojo-specific metadata associated with the instance.
    pub metadata: InstanceMetadata,

    /// Correpsonds to the Name property of the instance.
    pub name: Cow<'source, str>,

    /// Corresponds to the ClassName property of the instance.
    pub class_name: Cow<'source, str>,

    /// All other properties of the instance, weakly-typed.
    pub properties: HashMap<String, RbxValue>,

    /// The children of the instance represented as more snapshots.
    ///
    /// Order is relevant for Roblox instances!
    pub children: Vec<InstanceSnapshot<'source>>,
}

impl<'source> InstanceSnapshot<'source> {
    pub fn get_owned(&'source self) -> InstanceSnapshot<'static> {
        let children: Vec<InstanceSnapshot<'static>> = self
            .children
            .iter()
            .map(InstanceSnapshot::get_owned)
            .collect();

        InstanceSnapshot {
            snapshot_id: None,
            metadata: self.metadata.clone(),
            name: Cow::Owned(self.name.clone().into_owned()),
            class_name: Cow::Owned(self.class_name.clone().into_owned()),
            properties: self.properties.clone(),
            children,
        }
    }

    pub fn from_tree(tree: &RbxTree, id: RbxId) -> InstanceSnapshot<'static> {
        let instance = tree
            .get_instance(id)
            .expect("instance did not exist in tree");

        let children = instance
            .get_children_ids()
            .iter()
            .cloned()
            .map(|id| InstanceSnapshot::from_tree(tree, id))
            .collect();

        InstanceSnapshot {
            snapshot_id: Some(id),
            metadata: InstanceMetadata::default(),
            name: Cow::Owned(instance.name.clone()),
            class_name: Cow::Owned(instance.class_name.clone()),
            properties: instance.properties.clone(),
            children,
        }
    }
}

impl<'source> Default for InstanceSnapshot<'source> {
    fn default() -> InstanceSnapshot<'source> {
        InstanceSnapshot {
            snapshot_id: None,
            metadata: InstanceMetadata::default(),
            name: Cow::Borrowed("DEFAULT"),
            class_name: Cow::Borrowed("DEFAULT"),
            properties: HashMap::new(),
            children: Vec::new(),
        }
    }
}
