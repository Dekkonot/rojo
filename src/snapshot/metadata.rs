use std::path::PathBuf;

use crate::project::ProjectNode;

/// Rojo-specific metadata that can be associated with an instance or a snapshot
/// of an instance.
#[derive(Debug, Clone, PartialEq)]
pub struct InstanceMetadata {
    /// Whether instances not present in the source should be ignored when
    /// live-syncing. This is useful when there are instances that Rojo does not
    /// manage.
    pub ignore_unknown_instances: bool,

    /// A complete view of where this snapshot came from. It should contain
    /// enough information, if not None, to recreate this snapshot
    /// deterministically assuming the source has not changed state.
    pub source: Option<InstanceSource>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum InstanceSource {
    File {
        path: PathBuf,
    },
    ProjectFile {
        path: PathBuf,
        name: String,
        node: ProjectNode,
    },
}

impl Default for InstanceMetadata {
    fn default() -> Self {
        InstanceMetadata {
            ignore_unknown_instances: false,
            source: None,
        }
    }
}
