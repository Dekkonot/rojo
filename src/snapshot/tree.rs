use std::{collections::HashMap, path::PathBuf};

use rbx_dom_weak::{RbxId, RbxInstance, RbxInstanceProperties, RbxTree};

use crate::mapset::MapSet;

use super::InstanceMetadata;

/// An expanded variant of rbx_dom_weak's `RbxTree` that tracks additional
/// metadata per instance that's Rojo-specific.
///
/// This tree is also optimized for doing fast incremental updates and patches.
#[derive(Debug)]
pub struct RojoTree {
    /// Contains the instances without their Rojo-specific metadata.
    inner: RbxTree,

    /// Metadata associated with each instance that is kept up-to-date with the
    /// set of actual instances.
    metadata: HashMap<RbxId, InstanceMetadata>,

    /// A multimap from source paths to all of the root instances that were
    /// constructed from that path.
    ///
    /// Descendants of those instances should not be contained in the set, the
    /// value portion of the map is also a set in order to support the same path
    /// appearing multiple times in the same Rojo project. This is sometimes
    /// called "path aliasing" in various Rojo documentation.
    path_to_id: MapSet<PathBuf, RbxId>,
}

impl RojoTree {
    pub fn new(root: InstancePropertiesWithMeta) -> RojoTree {
        let inner = RbxTree::new(root.inner);
        let mut metadata = HashMap::new();
        metadata.insert(inner.get_root_id(), root.metadata);

        RojoTree {
            inner,
            metadata,
            path_to_id: MapSet::new(),
        }
    }

    pub fn get_root_id(&self) -> RbxId {
        self.inner.get_root_id()
    }

    pub fn get_instance(&self, id: RbxId) -> Option<InstanceWithMeta> {
        if let Some(inner) = self.inner.get_instance(id) {
            let metadata = self.metadata.get(&id).unwrap();

            Some(InstanceWithMeta { inner, metadata })
        } else {
            None
        }
    }

    pub fn get_instance_mut(&mut self, id: RbxId) -> Option<InstanceWithMetaMut> {
        if let Some(inner) = self.inner.get_instance_mut(id) {
            let metadata = self.metadata.get_mut(&id).unwrap();

            Some(InstanceWithMetaMut { inner, metadata })
        } else {
            None
        }
    }

    pub fn insert_instance(
        &mut self,
        properties: InstancePropertiesWithMeta,
        parent_id: RbxId,
    ) -> RbxId {
        let id = self.inner.insert_instance(properties.inner, parent_id);
        self.metadata.insert(id, properties.metadata);
        id
    }

    pub fn remove_instance(&mut self, id: RbxId) -> Option<RojoTree> {
        if let Some(inner) = self.inner.remove_instance(id) {
            let mut metadata = HashMap::new();
            let mut path_to_id = MapSet::new(); // TODO

            let root_meta = self.metadata.remove(&id).unwrap();

            metadata.insert(id, root_meta);

            for instance in inner.descendants(id) {
                let instance_meta = self.metadata.remove(&instance.get_id()).unwrap();
                metadata.insert(instance.get_id(), instance_meta);
            }

            Some(RojoTree {
                inner,
                metadata,
                path_to_id,
            })
        } else {
            None
        }
    }
}

#[derive(Debug, Clone)]
pub struct InstancePropertiesWithMeta {
    pub inner: RbxInstanceProperties,
    pub metadata: InstanceMetadata,
}

#[derive(Debug)]
pub struct InstanceWithMeta<'a> {
    pub inner: &'a RbxInstance,
    pub metadata: &'a InstanceMetadata,
}

#[derive(Debug)]
pub struct InstanceWithMetaMut<'a> {
    pub inner: &'a mut RbxInstance,
    pub metadata: &'a mut InstanceMetadata,
}
