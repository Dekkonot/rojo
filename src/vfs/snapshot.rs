// This file is non-critical and used for testing, so it's okay if it's unused.
#![allow(unused)]

use std::collections::HashMap;

#[derive(Debug, Clone)]
pub enum VfsSnapshot {
    File(FileSnapshot),
    Directory(DirectorySnapshot),
}

impl VfsSnapshot {
    /// Create a new file VfsSnapshot with the given contents.
    pub fn file(contents: impl Into<Vec<u8>>) -> VfsSnapshot {
        VfsSnapshot::File(FileSnapshot {
            contents: contents.into(),
        })
    }

    /// Create a new directory VfsSnapshot with the given children.
    pub fn dir<S: Into<String>>(children: HashMap<S, VfsSnapshot>) -> VfsSnapshot {
        let children = children.into_iter().map(|(k, v)| (k.into(), v)).collect();

        VfsSnapshot::Directory(DirectorySnapshot { children })
    }

    pub fn empty_dir() -> VfsSnapshot {
        VfsSnapshot::Directory(DirectorySnapshot {
            children: Default::default(),
        })
    }
}

#[derive(Debug, Clone)]
pub struct FileSnapshot {
    pub contents: Vec<u8>,
}

#[derive(Debug, Clone)]
pub struct DirectorySnapshot {
    pub children: HashMap<String, VfsSnapshot>,
}
