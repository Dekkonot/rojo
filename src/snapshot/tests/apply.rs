use std::collections::HashMap;

use insta::assert_yaml_snapshot;
use maplit::hashmap;
use rbx_dom_weak::{RbxId, RbxInstanceProperties, RbxValue};
use serde::Serialize;

use rojo_insta_ext::RedactionMap;

use crate::snapshot::{
    apply_patch_set, InstanceMetadata, InstancePropertiesWithMeta, PatchSet, PatchUpdate, RojoTree,
};

#[test]
fn set_name_and_class_name() {
    let mut redactions = RedactionMap::new();

    let mut tree = empty_tree();
    redactions.intern(tree.get_root_id());

    let patch_set = PatchSet {
        updated_instances: vec![PatchUpdate {
            id: tree.get_root_id(),
            changed_name: Some("Hello, world!".to_owned()),
            changed_class_name: Some("Folder".to_owned()),
            changed_properties: Default::default(),
            changed_metadata: None,
        }],
        ..Default::default()
    };

    let applied_patch_set = apply_patch_set(&mut tree, patch_set);

    let tree_view = view_tree(&tree);
    let tree_value = redactions.redacted_yaml(tree_view);
    assert_yaml_snapshot!(tree_value);

    let applied_patch_value = redactions.redacted_yaml(applied_patch_set);
    assert_yaml_snapshot!(applied_patch_value);
}

#[test]
fn add_property() {
    let mut redactions = RedactionMap::new();

    let mut tree = empty_tree();
    redactions.intern(tree.get_root_id());

    let patch_set = PatchSet {
        updated_instances: vec![PatchUpdate {
            id: tree.get_root_id(),
            changed_name: None,
            changed_class_name: None,
            changed_properties: hashmap! {
                "Foo".to_owned() => Some(RbxValue::String {
                    value: "Value of Foo".to_owned(),
                }),
            },
            changed_metadata: None,
        }],
        ..Default::default()
    };

    let applied_patch_set = apply_patch_set(&mut tree, patch_set);

    let tree_view = view_tree(&tree);
    let tree_value = redactions.redacted_yaml(tree_view);
    assert_yaml_snapshot!(tree_value);

    let applied_patch_value = redactions.redacted_yaml(applied_patch_set);
    assert_yaml_snapshot!(applied_patch_value);
}

#[test]
fn remove_property() {
    let mut redactions = RedactionMap::new();

    let mut tree = empty_tree();
    redactions.intern(tree.get_root_id());

    {
        let root_id = tree.get_root_id();
        let mut root_instance = tree.get_instance_mut(root_id).unwrap();

        root_instance.properties_mut().insert(
            "Foo".to_owned(),
            RbxValue::String {
                value: "Should be removed".to_owned(),
            },
        );
    }

    let tree_view = view_tree(&tree);
    let tree_value = redactions.redacted_yaml(tree_view);
    assert_yaml_snapshot!("remove_property_initial", tree_value);

    let patch_set = PatchSet {
        updated_instances: vec![PatchUpdate {
            id: tree.get_root_id(),
            changed_name: None,
            changed_class_name: None,
            changed_properties: hashmap! {
                "Foo".to_owned() => None,
            },
            changed_metadata: None,
        }],
        ..Default::default()
    };

    let applied_patch_set = apply_patch_set(&mut tree, patch_set);

    let tree_view = view_tree(&tree);
    let tree_value = redactions.redacted_yaml(tree_view);
    assert_yaml_snapshot!("remove_property_after_patch", tree_value);

    let applied_patch_value = redactions.redacted_yaml(applied_patch_set);
    assert_yaml_snapshot!("remove_property_appied_patch", applied_patch_value);
}

fn empty_tree() -> RojoTree {
    RojoTree::new(InstancePropertiesWithMeta {
        properties: RbxInstanceProperties {
            name: "ROOT".to_owned(),
            class_name: "ROOT".to_owned(),
            properties: Default::default(),
        },
        metadata: Default::default(),
    })
}

/// Copy of data from RojoTree in the right shape to have useful snapshots.
#[derive(Debug, Serialize)]
struct InstanceView {
    id: RbxId,
    name: String,
    class_name: String,
    properties: HashMap<String, RbxValue>,
    metadata: InstanceMetadata,
    children: Vec<InstanceView>,
}

fn view_tree(tree: &RojoTree) -> InstanceView {
    view_instance(tree, tree.get_root_id())
}

fn view_instance(tree: &RojoTree, id: RbxId) -> InstanceView {
    let instance = tree.get_instance(id).unwrap();

    InstanceView {
        id: instance.id(),
        name: instance.name().to_owned(),
        class_name: instance.class_name().to_owned(),
        properties: instance.properties().clone(),
        metadata: instance.metadata().clone(),
        children: instance
            .children()
            .iter()
            .copied()
            .map(|id| view_instance(tree, id))
            .collect(),
    }
}
