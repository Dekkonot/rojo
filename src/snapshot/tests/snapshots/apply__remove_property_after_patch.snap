---
source: src/snapshot/tests/apply.rs
expression: tree_value
---
id: id-1
name: ROOT
class_name: ROOT
properties: {}
metadata:
  ignore_unknown_instances: false
  source_path: ~
  project_node: ~
children: []
